### ISSUE TITLE
404 Error on local file system access

### ISSUE THREAD

**Issue by CxRes** [jeff-zucker/solid-file-client#130]
I get a 404 Error on trying access local file system with solid-file-client 1.0.2. Something like `fc.readFolder` or `fc.readFile` fails with `file:` calls but not with `https:`.  

I am not able to say where the error is as I cannot get the unbundled version to run because it mixes cjs and esm calls. I am also unable to run tests; the `TEST_PREFIX` thing stumps me - I have never used jest!

node: 12.16.1
os: Windows 10 x64

**Comment by CxRes** [comment]
I got the tests to work!

Strangely on `file://`,  only `tests/SolidApi.links.test.js` fails. I could provide you a stack dump if you like.

On `app://` its more of a mess. On `https://` I am not sure how to setup login... 

Again Windows is more of a pain in general (but where mom and pop users are).

**Comment by Otto-AA** [comment]
I've run it with travis in a windows environment now and it reproduced your errors with `file://` at `tests/SolidApi.links.test.js`.

The immediate cause of it is, that it expects relative paths, but gets absolute paths like here:
```
  ● copying links › copyAclFileForItem › modify absolute paths in acl file of a file to match new location agent:to_target

    expect(received).resolves.toEqual(expected) // deep equality

    - Expected  - 3
    + Received  + 3

      ↵
      @prefix : <#>.
      @prefix n0: <http://www.w3.org/ns/auth/acl#>.
      @prefix n1: <http://xmlns.com/foaf/0.1/>.
    - @prefix c: </profile/card#>.
    + @prefix c: <file://C:\Users\travis\build\Otto-AA\solid-file-client/profile/card#>.

      :ControlReadWrite
          a n0:Authorization;
    -     n0:accessTo <./target.txt>;
    +     n0:accessTo <file://C:\Users\travis\build\Otto-AA\solid-file-client/test-folder/SolidApi-links/copying-links/target.txt>;
          n0:accessTo <./target.txt>;
    -     n0:agent </profile/card#me>;
    +     n0:agent <file://C:\Users\travis\build\Otto-AA\solid-file-client/profile/card#me>;
          n0:agent c:me;
          n0:agentClass n1:Agent;
          n0:default <./>;
          n0:mode n0:Control, n0:Read, n0:Write.
```

So the problem is likely the `getRootUrl` method in SolidApi.links.test.js which uses `/`:
```javascript
const getRootUrl = url => {
  const base = url.split('/')
  let rootUrl = base[0]
  let j = 0
  for (let i = 1; i < base.length - 1; i++) {
    j = i
    if (base[i] === '') { rootUrl += '/' }
    break
  }
  rootUrl = rootUrl + '/' + base[j + 1] // + ('/')
  return rootUrl
}
```

If this is rewritten to work with windows and unix paths it maybe fixes the problem. Maybe there have to be other changes too to make it work for windows paths.

The method has been written by @bourgeoa , so he likely knows more about its function. To slightly simplify it a bit, you could replace the foor loop with an if, but it's still a bit cryptic to me:
```javscript
const getRootUrl = url => {
  const base = url.split('/')
  let rootUrl = base[0]
  let j = 0
  if (1 < base.length - 1) {
    j = 1
    if (base[1] === '')
      rootUrl += '/'
  }
  rootUrl = rootUrl + '/' + base[j + 1] // + ('/')
  return rootUrl
}
```

**Comment by Otto-AA** [comment]
Regarding `TEST_PREFIX=app://ls/` the problem is that apparently `app://ls/test-folder/` is not initialized and therefore it can't create the other test folders (resulting in much error logging). Not sure why this does not work.

**Comment by CxRes** [comment]
I had pushed some changes to solid-rest with regards to windows paths a while back. You might want to merge them. Those should fix a part of the Windows specific errors.

Also you may probably want to exclusively use the native node functions `url.fileURLToPath` and `url.pathToFileURL` to ensure all path-url conversions work cross-platform for local files.

**Comment by jeff-zucker** [comment]
The 404 errors are, I believe caused by non-existence of the top level folders.  I think we may need to create the top levels with PUT rather than POST.  Also, as @CxRes says, there are needed changes to solid-rest.  I am working on fixing Solid Rest but it's a bit complicated because I made a bunch of changes in a separate branch in order to get it to work with Data Kitchen.  Once I've re-integrated those and added @CxRes' patches, I'll set Travis on it and solid-auth-cli and I suspect fixing the Windows stuff in those will mostly fix it in solid-file-client.

**Comment by jeff-zucker** [comment]
Allright, I have fixed a number of things in solid-rest and now all of the solid-file-client app:// tests work on all 3 OSs.  The only fail is on the one mentioned above in SolidApi.links on Windows.  @CxRes - I have left your PR on solid-rest commented out in the code.  I can get all Travis tests to pass on all OSs with solid-rest the way I have it but will also look at your improvements.   I don't know if y'all can see the Travis results, if not, let me know and maybe there's a way I can provide access.
 https://travis-ci.com/github/jeff-zucker/solid-file-client/jobs/321874612

**Comment by CxRes** [comment]
@jeff-zucker I think the test fail not because changes to solid-rest are wrong, but that one needs to generate proper url from paths for the tests which require the following two changes: 

replace https://github.com/jeff-zucker/solid-file-client/blob/767179fa68258c80b491bd34d3126dfedeb6007b/tests/utils/FileApi.js#L15
with
```
return libUrl.fileURLToPath(url)
```

and 

replace  https://github.com/jeff-zucker/solid-file-client/blob/767179fa68258c80b491bd34d3126dfedeb6007b/tests/utils/contextSetup.js#L71
with
```
return libUrl.fileURLToPath(url)
```

Do not forget to import `libUrl` in both these files:
```
import libUrl from 'url'
```

I'll an also fixing solid-rest as you have requested. Try running the tests with changes above after I have done so!

**Comment by jeff-zucker** [comment]
@cxRes - did login work for you on windows in latest solid-auth-cli without the Assertion error?  It works for me but others are still getting that error.

**Comment by CxRes** [comment]
@jeff-zucker I had just copy pasted the changed code since npm was not delivering the changed code. But I had left my old fix inside solid-cli as is. I disabled that right now and it still works.

**Comment by jeff-zucker** [comment]
@cxRes, I implemented what you suggested above and it made the file tests fail on all three OSs.

**Comment by CxRes** [comment]
If tests fail on all three OSes I suspect that there is a syntax or semantic error. Again, all tests seem to pass for me except an occasional EPERM (which I believe is unrelated to the library and windows permissions specific). Could you post the error message?

**Comment by CxRes** [comment]
Also solid-rest did not update with my changes, so that could be the issue.

**Comment by jeff-zucker** [comment]
Sorry, don't understand what you mean by it did not update.

**Comment by jeff-zucker** [comment]
With version 1.1.7 of solid-rest, all tests pass for app and file in all three OSs.  These are without any of your changes.  Tomorrow I will try adding your changes in one at a time and see if the tests continue to pass.

**Comment by CxRes** [comment]
I meant that once I resolved the conflicts and tried to merge changes, Travis stopped the merge because the tests were not passing. 

Now the status is this afaict:
1) solid-rest fails with my changes for me as well (see my comment in solid-rest PR).
2) solid-file-client passes test with my changes to solid-rest and the two changes above to tests. here.

There is something I am missing here, I need to think...

**Comment by Otto-AA** [comment]
@CxRes 

The tests ([on travis](https://github.com/jeff-zucker/solid-file-client/runs/607263065)) currently only fail on windows with file:// and in this OS, only fail for tests/SolidApi.links.test.js.

I understand why you think that the paths in FileApi.js need to updated, but considering that all other tests pass, which frequently use FileApi.js themselves, I would be very surprised if this is the problem. Therefore I'd rather focus on what the differences between tests/SolidApi.links.test.js and the other tests are.

As suggested above, I would take a look at `getRootUrl` in tests/SolidApi.links.test.js first. If this doesn't resolve all issues, continue with acl and meta creation in TestFolderGenerator.js.

**Comment by CxRes** [comment]
@Otto-AA Thats because only Windows has really silly non-standard rules for file paths. I am now certain of what the problem is and have written a comment in solid-rest#22. The other tests are passing because the you can get node fs to work with posix paths even on windows, links are the only non-fs component. In fact, if you were to log the paths and urls generated on a Windows machine, you will find a mess of forward and back slashes, which confuses the link generation.

I am now able to get all tests to pass (with changes to tests here and changes in solid-rest), except some random EPERM errors on delete (again Windows madness that is independent of this issue, related to timing of the tests). In particular, all link tests pass and have well formed url's during the intermediate steps. The important thing is that all fixes need to be orchestrated together.

**Comment by jeff-zucker** [comment]
Well nuts.  The solid-rest and solid-auth-cli tests are passing everywhere and the solid-file-client tests pass everywhere except in Windows with the file protocol and now they are failing there in several tests, not just links.

**Comment by jeff-zucker** [comment]
I tried modifying the getRoots method as @Otto-AA suggested.  I replaced "/" with libPath.sep ... did not seem to fix the test.

**Comment by jeff-zucker** [comment]
Am I correct that this is currently an issue with how the tests are written and that as far as we know, when fed appropriate URIs, solid-file-client will behave correctly even on windows?

**Comment by jeff-zucker** [comment]
The links test appears to work now but the composed test does not.

**Comment by CxRes** [comment]
Have tried to run changes in FileApi.js together with ContextSetup.js. That I think did the trick for me...

**Comment by CxRes** [comment]
Also, imho you should not touch `getRootUrl` because url's are only supposed to have forward slashes.

**Comment by jeff-zucker** [comment]
Right, I reverted getRootUrl because it also fouled up app.  So currently I have your changes in FileApi and in ContextSetup but it isn't passing for me in Windows with file.

**Comment by Otto-AA** [comment]
> I am now able to get all tests to pass (with changes to tests here and changes in solid-rest), except some random EPERM errors on delete (again Windows madness that is independent of this issue, related to timing of the tests). In particular, all link tests pass and have well formed url's during the intermediate steps. The important thing is that all fixes need to be orchestrated together.

If you got this working, could you try to create a PR with your changes and a description of what else needs to be done from your point of view?

@jeff-zucker When you try out changes to fix this issue, it would probably be better to do them on a different branch. Commiting to master, the git log gets a bit spammed, merging stuff could be a bit harder and people who clone the repo get an inconsistent version. None of this is a big problem, but doing it on a separate branch has no downsides so I'd prefer that :)

**Comment by CxRes** [comment]
@Otto-AA I have made just the two changes I have mentioned above (though on the master that was valid until @jeff-zucker came back from his hiatus) and changes in solid-rest which I had PRed a few weeks back. A lot of changes have been made since then though I am still surprised that they would have side effects on this issue. I'll do my best to help though!

@jeff-zucker I second @Otto-AA's suggestion to make changes on a branch. Could you also please squash the changes here, so that those little experiments are suppressed? Then I'll try to see if I can get it to work on a new branch!

**Comment by jeff-zucker** [comment]
Sorry for not creating a new branch.  I have so many $$*_)7-097 packages, I am loathe to create new folders.  Will do it now.

@CxRes not sure what you mean by squash the changes, AFAIK, I have merged everything you've committed and put in by hand your other suggestions.  They are in the current, master .  

Both: is this a problem with the tests or is the package misbehaving outside the test environement?

**Comment by jeff-zucker** [comment]
Also: I npm'd the latest master because of the slug-on-PUT bug.

**Comment by jeff-zucker** [comment]
The new branch : https://github.com/jeff-zucker/solid-file-client/tree/cxRes-changes

**Comment by Otto-AA** [comment]
> Both: is this a problem with the tests or is the package misbehaving outside the test environement?

I don't know if the library itself is working in windows, or what bugs might exist there. I guess that the test fails are because the test environment is not windows-friendly, not because of the solid-file-client library itself. But I'm not sure about that.

**Comment by Otto-AA** [comment]
Is everything here fixed when #147 is merged?

**Comment by jeff-zucker** [comment]
Yes, all Travis tests pass after #147.  Yay!  Thanks all.
